tests/functional/lib/BackbeatConsumer.js

```diff
@@ -0,0 +1,221 @@
+const assert = require('assert');
+const async = require('async');
+const BackbeatProducer = require('../../../lib/BackbeatProducer');
+const BackbeatConsumer = require('../../../lib/BackbeatConsumer');
+const zookeeperConf = { connectionString: 'localhost:2181' };
+const kafkaConf = { hosts: 'localhost:9092' };
+
+describe('BackbeatConsumer concurrency tests', () => {
+    const topicConc = 'backbeat-consumer-spec-conc-1000';
+    const groupIdConc = `replication-group-conc-${Math.random()}`;
+    let producer;
+    let consumer;
+    let consumedMessages = [];
+
+    function queueProcessor(message, cb) {
+        if (message.value.toString() !== 'taskStuck') {
+            consumedMessages.push(message.value);
+            process.nextTick(cb);
+        }
+    }
+    before(function before(done) {
+        this.timeout(60000);
+
+        producer = new BackbeatProducer({
+            kafka: kafkaConf,
+            topic: topicConc,
+            pollIntervalMs: 100,
+        });
+        consumer = new BackbeatConsumer({
+            zookeeper: zookeeperConf,
+            kafka: kafkaConf, groupId: groupIdConc, topic: topicConc,
+            queueProcessor,
+            concurrency: 10,
+            bootstrap: true,
+        });
+        async.parallel([
+            innerDone => producer.on('ready', innerDone),
+            innerDone => consumer.on('ready', innerDone),
+        ], done);
+    });
+    afterEach(() => {
+        consumedMessages = [];
+        consumer.removeAllListeners('consumed');
+    });
+    after(function after(done) {
+        this.timeout(10000);
+        async.parallel([
+            innerDone => producer.close(innerDone),
+            innerDone => consumer.close(innerDone),
+        ], done);
+    });
+
+    it('should be able to process 1000 messages with concurrency', done => {
+        const boatloadOfMessages = [];
+        for (let i = 0; i < 1000; ++i) {
+            boatloadOfMessages.push({
+                key: `message-${i}`,
+                message: `{"message_index":"${i}"}`,
+            });
+        }
+        async.series([
+            next => {
+                setTimeout(() => producer.send(boatloadOfMessages, err => {
+                    assert.ifError(err);
```

Comment:
This is asynchronous and the code following this i synchronous, I think it deserves its own step.

Reply:
The synchronous code executes immediately without waiting for`producer.send` to complete, this is done on purpose so that we produce messages only after we are ready to consume them. If you have another preference for the code style let me know.